In [1]:
import sklearn as sk
from sklearn import preprocessing
import math
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
#import mglearn
import os
import inspect
import sys


sys.path.append('./../../')

# OPTIONAL: Load the "autoreload" extension so that code can change
%load_ext autoreload

#Own Library modules
import src.pca

# OPTIONAL: always reload modules so that as you change code in src, it gets loaded
%autoreload 2

from src.pca import PCA_from_sklearn

#### Objetivo:

El objetivo de este notebook es utilizar PCA de scikit-learn para la reducción de dimensionalidad por sección, ya que existen más de 150 variables.


In [2]:
df = pd.read_csv('datos_limpios.csv', encoding = "L1")

##### Sección 1. Datos generales

In [3]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_1 = df.loc[:, ['p1', 'p2', 'p3_A', 'p3_B','p3_C','p4_A', 'p4_B','p4_C','p5_A', 'p5_B','p5_C']] ###Se seleccionan las preguntas que conforman esta sección

In [4]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_1 = scaler.fit_transform(seccion_1)

In [5]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_1)

In [6]:
componentes = pd.DataFrame(pca.components_.T, index=['p1', 'p2', 'p3_A', 'p3_B','p3_C','p4_A', 'p4_B','p4_C','p5_A', 'p5_B','p5_C'])

In [7]:
####Se eligieron tres componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos 
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2', 2:'principal_component_3'})
componentes

principal_component_1  principal_component_2  principal_component_3
p1                 0.032745              -0.027338               0.038322
p2                 0.042043               0.120472              -0.053366
p3_A              -0.038879              -0.694961               0.092536
p3_B               0.022584               0.658153              -0.091274
p3_C               0.032970               0.217248              -0.009455
p4_A               0.443411              -0.070732              -0.335598
p4_B              -0.511101               0.019803              -0.013504
p4_C               0.223747               0.071486               0.609995
p5_A               0.431409              -0.078579              -0.345323
p5_B              -0.498698               0.017791              -0.046291
p5_C               0.226581               0.066405               0.610252

In [8]:
varianza_explicada

array([0.30080223, 0.1789151 , 0.15889631])

Para el primer componente principal se asocia la pregunta 4 respuesta b, para el segundo componente principal se asocia la pregunta 3 respuesta a y para el tercer componente principal se asocia la pregunta 5 respuesta c. El porcentaje de varianza explicada para cada uno de los componentes es de 30.08%, 17.89% y 15.89%, respectivamente.  

- Como se puede ver a continuación, los tres eigenvalues son mayores a 1, por lo tanto se eligen tres componentes para la regresión.

In [9]:
eigenvalues

array([3.30883043, 1.96806964, 1.74786252])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [10]:
componentesprincipales_seccion_1 = pd.DataFrame(componentesprincipales, columns=['sec_1_cp_1', 'sec_1_cp_2', 'sec_1_cp_3'])
componentesprincipales_seccion_1

sec_1_cp_1  sec_1_cp_2  sec_1_cp_3
0        -0.821715   -0.066756   -0.106842
1        -0.856175   -0.408062    0.096823
2        -0.921727   -0.353333    0.020105
3        -0.856175   -0.408062    0.096823
4        -0.856175   -0.408062    0.096823
...            ...         ...         ...
560621   -0.871721   -0.210044   -0.043369
560622   -0.806169   -0.264774    0.033350
560623   -0.806169   -0.264774    0.033350
560624   -0.871721   -0.210044   -0.043369
560625   -0.921727   -0.353333    0.020105

[560626 rows x 3 columns]

##### Sección 2. Situación laboral 

In [11]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_2 = df.loc[:, ['p6', 'p7', 'p8', 'p9_A', 'p9_B', 'p9_C', 'p9_D', 'p9_E', 'p9_F', 'p9_G']] ###Se seleccionan las preguntas que conforman esta sección

In [12]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_2 = scaler.fit_transform(seccion_2)

In [13]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_2)

In [14]:
componentes = pd.DataFrame(pca.components_.T, index=['p6', 'p7', 'p8', 'p9_A', 'p9_B', 'p9_C', 'p9_D', 'p9_E', 'p9_F', 'p9_G'])

In [15]:
####Se eligieron tres componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos 
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2', 2:'principal_component_3'})
componentes

principal_component_1  principal_component_2  principal_component_3
p6                 0.474838              -0.031406               0.025848
p7                 0.461136               0.018184               0.089782
p8                 0.407272               0.063086              -0.109050
p9_A              -0.484587               0.027653               0.001719
p9_B               0.070410               0.028610               0.059394
p9_C               0.237947              -0.744066              -0.300836
p9_D               0.222406               0.653731              -0.438945
p9_E               0.104603               0.039955               0.121175
p9_F               0.105717               0.074345               0.060952
p9_G               0.167421               0.070693               0.821143

In [16]:
varianza_explicada

array([0.39730487, 0.11419303, 0.10984288])

Para el primer componente principal la pregunta 9 respuesta a, es la que más se asocia al primer componente principal, mientras que la pregunta 9 respuesta c, se asocia al segundo componente principal y la pregunta 9 respuesta g se asocia al tercer componente principal. El porcentaje de varianza explicada de cada componente es de 39.73%, 11.42% y de 10.98% respectivamente.  

- Los eigenvalues que a continuación se presentan son mayores a 1, por lo que con este criterio se aceptan los tres componentes que serán utilizados en la regresión.

In [17]:
eigenvalues

array([3.97305579, 1.14193229, 1.09843076])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [18]:
componentesprincipales_seccion_2 = pd.DataFrame(componentesprincipales, columns=['sec_2_cp_1', 'sec_2_cp_2', 'sec_2_cp_3'])
componentesprincipales_seccion_2

sec_2_cp_1  sec_2_cp_2  sec_2_cp_3
0        -1.544900    0.025678    0.000465
1        -1.544900    0.025678    0.000465
2         3.215723   -2.008007   -1.164138
3        -1.544900    0.025678    0.000465
4        -1.544900    0.025678    0.000465
...            ...         ...         ...
560621   -1.303537    0.035195    0.047458
560622   -1.544900    0.025678    0.000465
560623    2.698132    0.301093    3.615794
560624    2.456769    0.291575    3.568801
560625   -1.544900    0.025678    0.000465

[560626 rows x 3 columns]

##### Sección 3. Expectativas educativas

In [19]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_3 = df.loc[:, ['p10', 'p11', 'p12', 'p13', 'p14']] ###Se seleccionan las preguntas que conforman esta sección

In [20]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_3 = scaler.fit_transform(seccion_3)

In [21]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_3)

In [22]:
componentes = pd.DataFrame(pca.components_.T, index=['p10', 'p11', 
                                                         'p12', 'p13',
                                                         'p14', 
                                                         ])

In [23]:
####Se eligieron dos componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos 
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2'})
componentes

principal_component_1  principal_component_2
p10               0.118769              -0.745795
p11               0.433563               0.450160
p12               0.487973               0.328743
p13               0.530490              -0.199727
p14               0.527617              -0.305260

In [24]:
####Se obtiene la varianza explicada por cada componente principal
varianza_explicada

array([0.49697126, 0.21837784])

Por lo tanto, las preguntas 13 y 14 se encuentran asociadas al primer componente principal a un porcentaje de varianza explicada de 49.70%, mientras que la pregunta 11 se asocia al componente principal 2 a un porcentaje de varianza explicada de 21.84%.

- Como se observa los eigenvalues son mayores a 1, siendo está otra medida para escoger el número de componentes, reduciendo así la dimensionalidad.

In [25]:
eigenvalues

array([2.48486074, 1.09189113])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [26]:
componentesprincipales_seccion_3 = pd.DataFrame(componentesprincipales, columns=['sec_3_cp_1', 'sec_3_cp_2'])
componentesprincipales_seccion_3

sec_3_cp_1  sec_3_cp_2
0        -2.751152    1.038033
1         1.439305   -0.274336
2        -0.075647   -1.163561
3        -2.054591    0.706529
4        -0.592569    1.411202
...            ...         ...
560621   -1.369152    0.309960
560622   -1.242848   -0.348250
560623    1.372021   -1.269330
560624   -1.481507   -0.938306
560625   -0.546287   -0.679755

[560626 rows x 2 columns]

##### Sección 4. Características personales

In [27]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados
seccion_4 = df.loc[:, ['p15', 'p16', 'p17', 'p18', 'p19', 'p20', 'p21', 'p22', 'p23', 'p24', 'p25', 'p26', 'p27',
                         'p28', 'p29', 'p30', 'p31', 'p32', 'p33', 'p34', 'p35', 'p36', 'p37', 'p38', 'p39', 'p40',
                         'p41', 'p42', 'p43', 'p44', 'p45', 'p46', 'p47', 'p48', 'p49', 'p50','p51', 'p52', 'p53', 
                         'p54', 'p55', 'p56']] ###Se seleccionan las preguntas que conforman esta sección

In [28]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_4 = scaler.fit_transform(seccion_4)

In [29]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_4)

In [30]:
componentes = pd.DataFrame(pca.components_.T, index=['p15', 'p16', 'p17', 'p18', 'p19', 'p20', 'p21', 'p22', 'p23', 'p24', 'p25', 'p26', 'p27',
                         'p28', 'p29', 'p30', 'p31', 'p32', 'p33', 'p34', 'p35', 'p36', 'p37', 'p38', 'p39', 'p40',
                         'p41', 'p42', 'p43', 'p44', 'p45', 'p46', 'p47', 'p48', 'p49', 'p50','p51', 'p52', 'p53', 
                         'p54', 'p55', 'p56'])

In [31]:
####Se eligieron quince componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2', 2:'principal_component_3', 3:'principal_component_4',
                                         4:'principal_component_5', 5:'principal_component_6', 6:'principal_component_7', 7:'principal_component_8', 8:'principal_component_9',
                                         9:'principal_component_10', 10:'principal_component_11', 11:'principal_component_12', 12:'principal_component_13', 13:'principal_component_14',
                                         14:'principal_component_15'})
componentes

principal_component_1  principal_component_2  principal_component_3  \
p15               0.167210              -0.055927              -0.027942   
p16              -0.034948              -0.146095               0.055253   
p17               0.190844               0.012932               0.023252   
p18              -0.110435              -0.246666               0.039936   
p19               0.132062              -0.079056               0.033859   
p20               0.181588              -0.053154               0.023434   
p21              -0.148552              -0.207479               0.075928   
p22              -0.119501              -0.241581               0.069853   
p23              -0.068941              -0.251757              -0.053505   
p24              -0.049889              -0.232934              -0.111155   
p25              -0.023720              -0.240038              -0.107043   
p26              -0.040946              -0.216907              -0.076644   
p27              -0.033632              -0.228087              -0.169550   
p28              -0.025439              -0.170606              -0.043272   
p29              -0.045206              -0.224374              -0.059713   
p30               0.085226               0.008521              -0.019911   
p31              -0.046747              -0.204449              -0.163307   
p32               0.091881              -0.030254              -0.054748   
p33               0.067740              -0.138190              -0.140340   
p34               0.184313              -0.111359              -0.281649   
p35               0.233939              -0.066390              -0.237486   
p36               0.234451              -0.050308              -0.222120   
p37               0.223897              -0.082662              -0.245504   
p38               0.233692              -0.059968              -0.225386   
p39               0.148125              -0.105071              -0.196203   
p40               0.117143              -0.158786              -0.110095   
p41              -0.109961              -0.274931               0.160751   
p42              -0.095425              -0.276621               0.171534   
p43              -0.095812              -0.246012               0.165293   
p44              -0.128802              -0.239287               0.140001   
p45               0.079558              -0.099285               0.186669   
p46               0.177672              -0.020716               0.283016   
p47               0.199820              -0.024794               0.223024   
p48               0.155806              -0.051983               0.319551   
p49               0.220206              -0.060999               0.223700   
p50               0.222983              -0.075380               0.193664   
p51               0.250260              -0.021685               0.152996   
p52               0.192974              -0.016878               0.083698   
p53               0.240654              -0.080924               0.095391   
p54               0.228606              -0.058593               0.104721   
p55               0.218847              -0.081089               0.088067   
p56               0.101824              -0.110183               0.060675   

     principal_component_4  principal_component_5  principal_component_6  \
p15               0.254246              -0.216400               0.048560   
p16               0.177005              -0.242452              -0.069229   
p17               0.280721              -0.177948               0.067674   
p18              -0.051170              -0.324875              -0.174267   
p19               0.331012              -0.328524               0.089998   
p20               0.319088              -0.288709               0.084329   
p21              -0.093686              -0.286803              -0.081413   
p22              -0.051865              -0.344112              -0.137676   
p23               0.167876               0.098631   

In [32]:
##Se convierte a valor absoluto el dataframe para poder explicar qué preguntas se asocian a cada componente.
componentes = componentes.abs()

In [33]:
componentes.idxmax(axis=0) ##Se obtienen las preguntas que en valor absoluto se asocian más a cada componente.

principal_component_1     p51
principal_component_2     p42
principal_component_3     p48
principal_component_4     p19
principal_component_5     p22
principal_component_6     p32
principal_component_7     p46
principal_component_8     p32
principal_component_9     p33
principal_component_10    p45
principal_component_11    p16
principal_component_12    p56
principal_component_13    p28
principal_component_14    p56
principal_component_15    p39
dtype: object

In [34]:
####Se obtiene la varianza explicada por cada componente principal
varianza_explicada = pd.DataFrame(varianza_explicada, index = componentes.columns)
varianza_explicada['var_exp'] = varianza_explicada[0] * 100
varianza_explicada = varianza_explicada.rename(columns = {'var_exp':'porcentaje_varianza_explicada'})
varianza_explicada = varianza_explicada.drop([0], axis = 1)

In [35]:
varianza_explicada ##Porcentaje de varianza explicada por cada componente principal

porcentaje_varianza_explicada
principal_component_1                       15.280862
principal_component_2                       10.562898
principal_component_3                        4.544976
principal_component_4                        3.788139
principal_component_5                        3.642231
principal_component_6                        2.898986
principal_component_7                        2.882884
principal_component_8                        2.842565
principal_component_9                        2.491572
principal_component_10                       2.401837
principal_component_11                       2.152930
principal_component_12                       2.069194
principal_component_13                       2.015835
principal_component_14                       1.982375
principal_component_15                       1.923326

- Aquí se puede observar que no todos los eigenvalues son mayores a 1, por lo que esta medida nos dice que nos quedemos con 10 componentes y son los que usaremos en la regresión.

In [36]:
eigenvalues

array([6.41797329, 4.43642522, 1.90889333, 1.59102105, 1.52973961,
       1.21757626, 1.21081334, 1.1938795 , 1.04646215, 1.00877321,
       0.9042323 , 0.86906304, 0.84665211, 0.83259913, 0.8077982 ])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [37]:
componentesprincipales_seccion_4_1 = pd.DataFrame(componentesprincipales)
componentesprincipales_seccion_4_1_regress = componentesprincipales_seccion_4_1.loc[:, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]]

In [38]:
componentesprincipales_seccion_4_1_regress = componentesprincipales_seccion_4_1_regress.rename(columns = {0:'sec_4_cp_1', 1:'sec_4_cp_2', 2:'sec_4_cp_3', 3:'sec_4_cp_4',
                                                                                                         4:'sec_4_cp_5', 5:'sec_4_cp_6', 6:'sec_4_cp_7', 
                                                                                                         7:'sec_4_cp_8', 8:'sec_4_cp_9', 9:'sec_4_cp_10'})

In [39]:
componentesprincipales_seccion_4_1_regress

sec_4_cp_1  sec_4_cp_2  sec_4_cp_3  sec_4_cp_4  sec_4_cp_5  \
0        -0.593721    1.570301    0.528639   -1.548947   -1.283364   
1        -0.678143    0.180167    1.529614    1.368652   -0.388323   
2         1.090394   -0.054025   -0.380154    0.248355   -0.996565   
3        -0.710868    3.044078    1.249950    0.945428   -1.383257   
4         1.208648    1.137097    0.517063   -0.031776    0.612268   
...            ...         ...         ...         ...         ...   
560621    3.742078    1.154400   -0.565152   -0.709642    1.073323   
560622    0.712407    2.776176    0.165692    0.055128   -0.308646   
560623   -3.329364   -3.384182    2.121597    1.845967   -0.147357   
560624    1.570937   -1.904671   -2.563952    1.018856   -1.175766   
560625   -0.083938   -2.960756   -0.857424    1.179406   -1.862259   

        sec_4_cp_6  sec_4_cp_7  sec_4_cp_8  sec_4_cp_9  sec_4_cp_10  
0         0.974512   -0.708282   -0.422397   -1.071882    -0.473350  
1         0.197635   -0.577255   -0.482915    2.123425     0.621808  
2        -0.910941    1.760915    0.085560    0.023151    -0.547788  
3        -0.178858    0.211006   -0.940099    1.427071     0.584912  
4         0.896264    0.179634   -0.852127    0.909318    -0.447715  
...            ...         ...         ...         ...          ...  
560621    0.147903    0.761750    0.029910    1.426307     1.419391  
560622    0.254449   -0.237424    0.777261   -0.643161    -0.738650  
560623   -1.657316    1.036598   -1.440583    0.990399    -0.213448  
560624   -1.429768   -0.742217   -0.952739   -1.435483     0.713499  
560625    1.027375    2.137302   -0.833068    1.111688     0.156951  

[560626 rows x 10 columns]

##### Sección 5. Inglés y cómputo.

- 5.1. Califica tu nivel de habilidad para leer en inglés.

In [40]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_5_1 = df.loc[:, ['p57', 'p58', 'p59', 'p60']] ###Se seleccionan las preguntas que conforman esta sección

In [41]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_5_1 = scaler.fit_transform(seccion_5_1)

In [42]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_5_1)

In [43]:
componentes = pd.DataFrame(pca.components_.T, index=['p57', 'p58', 'p59', 'p60'])

In [44]:
####Se eligió un componente principal, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2'})
componentes

principal_component_1
p57               0.506117
p58               0.517654
p59               0.488779
p60               0.486801

In [45]:
varianza_explicada

array([0.74731825])

Para el primer componente principal se asocian la preguntas 57 y 58. El porcentaje de varianza explicada para el primer componentes es de 74.73%.

- El eigenvalue asociado a este componente principal es mayor a 1 y es con el único componente principal que usamos para la regresión.

In [46]:
eigenvalues

array([2.98927833])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [47]:
componentesprincipales_seccion_5_1 = pd.DataFrame(componentesprincipales, columns=['sec_5_1_cp_1'])
componentesprincipales_seccion_5_1

sec_5_1_cp_1
0          -3.012403
1           1.329468
2          -1.121389
3          -1.763936
4          -0.536424
...              ...
560621     -2.402316
560622     -0.536424
560623      3.805447
560624      1.967849
560625     -2.402316

[560626 rows x 1 columns]

- 5.2. Con qué frecuencia realizas las siguientes actividades en la computadora

In [48]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_5_2 = df.loc[:, ['p61', 'p62', 'p63', 'p64']] ###Se seleccionan las preguntas que conforman esta sección

In [49]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_5_2 = scaler.fit_transform(seccion_5_2)

In [50]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_5_2)

In [51]:
componentes = pd.DataFrame(pca.components_.T, index=['p61', 'p62', 'p63', 'p64'])

In [52]:
####Se eligieron dos componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2'})
componentes

principal_component_1  principal_component_2
p61               0.450106              -0.719827
p62               0.548978              -0.023552
p63               0.536229               0.038504
p64               0.456602               0.692684

In [53]:
varianza_explicada

array([0.59243727, 0.19244308])

Para el primer componente principal se asocia la pregunta 62 y 63, mientras que para el segundo componente principal se asocia la pregunta 61. El porcentaje de varianza explicada para el primer componentes es de 59.24% y para el segundo componente principal es de 19.24%.

- Se observa que únicamente el primer eigenvalue asociado al primer componente principal es mayor a 1 y ese componente es el que se usa en la regresión.

In [54]:
eigenvalues

array([2.36975333, 0.76977369])

- Se obtiene los componentes principales que son los que se meteran a la regresión.

In [55]:
componentesprincipales_seccion_5_2 = pd.DataFrame(componentesprincipales)
componentesprincipales_seccion_5_2_regress = componentesprincipales_seccion_5_2.loc[:, [0]]

In [56]:
componentesprincipales_seccion_5_2_regress = componentesprincipales_seccion_5_2_regress.rename(columns={0:'sec_5_2_cp_1'})

In [57]:
componentesprincipales_seccion_5_2_regress

sec_5_2_cp_1
0          -0.686659
1          -0.710620
2          -0.596293
3          -0.696755
4          -2.404186
...              ...
560621     -1.461989
560622     -1.022208
560623     -0.596293
560624      0.785685
560625      0.245442

[560626 rows x 1 columns]

##### Sección 6. Actividades deportivas y culturales

In [58]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_6 = df.loc[:, ['p65', 'p66']] ###Se seleccionan las preguntas que conforman esta sección

In [59]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_6 = scaler.fit_transform(seccion_6)

In [60]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_6)

In [61]:
componentes = pd.DataFrame(pca.components_.T, index=['p65', 'p66'])

In [62]:
####Se eligieron dos componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2'})
componentes

principal_component_1  principal_component_2
p65               0.707107              -0.707107
p66               0.707107               0.707107

In [63]:
varianza_explicada

array([0.57688468, 0.42311532])

Ambas preguntas se asocian a los dos componentes principales con un porcentaje de varianza explicada para cada uno de 57.69% y de 42.31%, respectivamente.

- Como se puede ver a continuación, sólo un eigenvalue es mayor a 1, por lo que se elige el primer componente principal para incoporarlo a la regresión.

In [64]:
eigenvalues

array([1.15377141, 0.84623215])

- Se obtiene los componentes principales que son los que se meteran a la regresión.

In [65]:
componentesprincipales_seccion_6 = pd.DataFrame(componentesprincipales)
componentesprincipales_seccion_6_regress = componentesprincipales_seccion_6.loc[:, [0]]

In [66]:
componentesprincipales_seccion_6_regress = componentesprincipales_seccion_6_regress.rename(columns={0:'sec_6_cp_1'})

In [67]:
componentesprincipales_seccion_6_regress

sec_6_cp_1
0        -1.720060
1        -0.076522
2        -0.405230
3        -0.405230
4        -1.062645
...            ...
560621    0.175530
560622    0.884049
560623    1.844619
560624    0.226634
560625   -1.720060

[560626 rows x 1 columns]

##### Sección 7. Trayectoria académica

- 7.1. Primaria y secundaria

In [68]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_7_1 = df.loc[:, ['p67_A','p67_B', 'p67_C', 'p67_D', 'p67_E', 'p68', 'p69', 'p70_A', 'p70_B',
                         'p70_C', 'p70_D', 'p70_E', 'p70_F', 'p70_G','p70_H', 'p71']] ###Se seleccionan las preguntas que conforman esta sección

In [69]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_7_1 = scaler.fit_transform(seccion_7_1)

In [70]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_7_1)

In [71]:
componentes = pd.DataFrame(pca.components_.T, index=['p67_A','p67_B', 'p67_C', 'p67_D', 'p67_E', 'p68', 'p69', 'p70_A', 'p70_B',
                         'p70_C', 'p70_D', 'p70_E', 'p70_F', 'p70_G','p70_H', 'p71'])

In [72]:
####Se eligieron siete componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2', 2:'principal_component_3', 
                                         3:'principal_component_4', 4:'principal_component_5', 5:'principal_component_6', 6:'principal_component_7'})
componentes

principal_component_1  principal_component_2  principal_component_3  \
p67_A               0.154404              -0.629017               0.102081   
p67_B              -0.359742               0.351795              -0.251554   
p67_C               0.394885               0.387867               0.066922   
p67_D              -0.089399               0.064028               0.086789   
p67_E              -0.126294               0.087910               0.154720   
p68                 0.274108              -0.002564              -0.595284   
p69                -0.402452              -0.356173              -0.100151   
p70_A               0.453977               0.035467               0.231379   
p70_B              -0.125401              -0.319740              -0.273707   
p70_C              -0.404542               0.289716              -0.024107   
p70_D              -0.014181               0.012791               0.085925   
p70_E              -0.012915              -0.017981               0.056877   
p70_F              -0.006068              -0.006559               0.038557   
p70_G               0.012340               0.021113              -0.005411   
p70_H              -0.039524               0.043117               0.072537   
p71                 0.205358               0.034381              -0.616059   

       principal_component_4  principal_component_5  principal_component_6  \
p67_A              -0.135074               0.020544               0.085923   
p67_B              -0.193882              -0.254126              -0.308647   
p67_C               0.345694              -0.089840              -0.018961   
p67_D               0.114508               0.007205               0.822204   
p67_E               0.135505               0.670821              -0.084301   
p68                -0.090271               0.176077               0.105611   
p69                -0.286163               0.075042              -0.024174   
p70_A              -0.464720              -0.006654              -0.099114   
p70_B               0.634391              -0.158473              -0.148577   
p70_C              -0.189233               0.073600               0.258185   
p70_D               0.091232              -0.031592               0.075825   
p70_E               0.044328               0.005488               0.112742   
p70_F               0.035663              -0.010968               0.062443   
p70_G               0.058109               0.011114               0.125729   
p70_H               0.122489               0.601542              -0.204857   
p71                -0.137776               0.215733               0.153096   

       principal_component_7  
p67_A               0.016510  
p67_B               0.036554  
p67_C               0.005808  
p67_D              -0.186960  
p67_E               0.000663  
p68                 0.017390  
p69                -0.008856  
p70_A              -0.076493  
p70_B              -0.086003  
p70_C              -0.018951  
p70_D              -0.018908  
p70_E               0.952238  
p70_F               0.139559  
p70_G               0.139837  
p70_H              -0.034103  
p71                 0.045895

In [73]:
varianza_explicada

array([0.15069025, 0.12333791, 0.09720801, 0.08416729, 0.0756832 ,
       0.06598655, 0.06303318])

Para el primer componente principal se asocia la pregunta 70 respuesta a, para el segundo componente principal se asocia la pregunta 67 respuesta a, la pregunta 71 se asocia al tercer componente principal, la pregunta 70 respuesta b se asocia al cuarto componente principal, la pregunta 67 respuesta e se asocia al quinto componente principal, la pregunta 67 respuesta d se asocia al sexto componente principal, la pregunta 70 respuesta e se asocia al séptimo componente principal. El porcentaje de varianza explicada para cada uno de los componentes principales es de 15.07%, 12.33%, 9.72%, 8.42%, 7.57%, 6.60% y 6.30%, respectivamente.

- Se observa que los siete eigenvalues asociados a los componentes principales son mayores a 1, así que los siete componentes principales son los que se usan en la regresión.

In [74]:
eigenvalues

array([2.41104823, 1.97341014, 1.55533093, 1.34667906, 1.21093331,
       1.05578662, 1.00853267])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [75]:
componentesprincipales_seccion_7_1 = pd.DataFrame(componentesprincipales, columns=['sec_7_1_cp_1', 'sec_7_1_cp_2', 'sec_7_1_cp_3', 'sec_7_1_cp_4', 'sec_7_1_cp_5', 
                                                                                   'sec_7_1_cp_6', 'sec_7_1_cp_7'])
componentesprincipales_seccion_7_1

sec_7_1_cp_1  sec_7_1_cp_2  sec_7_1_cp_3  sec_7_1_cp_4  sec_7_1_cp_5  \
0           0.685316     -1.085846      1.122070     -0.990718     -0.059627   
1          -0.964696     -1.920554      1.271060      1.612597     -0.811177   
2          -0.009050     -1.855260     -1.152122      1.154437     -0.025908   
3           0.048219     -1.129375      2.737524     -0.685278     -0.583140   
4           1.003865     -1.064081      0.314343     -1.143438      0.202129   
...              ...           ...           ...           ...           ...   
560621     -0.708858     -0.433862     -0.595613     -0.730150      0.508751   
560622     -0.887430     -0.432192     -0.207807     -0.671342      0.394044   
560623      3.483890      2.520807     -0.179802      0.931349     -0.087974   
560624      0.965271     -1.038976      0.282228     -1.178542      0.234470   
560625     -1.006249      0.870537      2.017078     -0.802673     -1.210565   

        sec_7_1_cp_6  sec_7_1_cp_7  
0          -0.090809     -0.113142  
1          -0.503549     -0.226994  
2           0.015919     -0.099159  
3          -0.437121     -0.198365  
4           0.082347     -0.070530  
...              ...           ...  
560621      0.976219      0.048101  
560622      0.907417      0.036772  
560623      0.099973     -0.015494  
560624      0.117899     -0.050576  
560625     -1.317378     -0.171652  

[560626 rows x 7 columns]

- 7.2. Educación media superior (Bachillerato)

In [76]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_7_2 = df.loc[:, ['p72', 'p73', 'p74_A','p74_B','p74_C','p74_D', 'p74_E', 'p74_F', 'p75', 'p76', 'p77', 'p78']] ###Se seleccionan las preguntas que conforman esta sección

In [77]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_7_2 = scaler.fit_transform(seccion_7_2)

In [78]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_7_2)

In [79]:
componentes = pd.DataFrame(pca.components_.T, index=['p72', 'p73', 'p74_A','p74_B','p74_C','p74_D', 'p74_E', 'p74_F', 'p75', 'p76', 'p77', 'p78'])

In [80]:
####Se eligieron seis componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2', 2:'principal_component_3', 3:'principal_component_4',
                                         4:'principal_component_5', 5:'principal_component_6'})
componentes

principal_component_1  principal_component_2  principal_component_3  \
p72                 0.037595              -0.139872              -0.347077   
p73                -0.199201              -0.223587              -0.516092   
p74_A              -0.733016               0.296417              -0.028602   
p74_B               0.291908              -0.052775              -0.208639   
p74_C               0.277993              -0.086606              -0.074555   
p74_D               0.291083              -0.188858               0.022432   
p74_E               0.119365              -0.043069               0.097509   
p74_F               0.108756              -0.058563               0.365459   
p75                 0.249669               0.662655              -0.102833   
p76                -0.026447               0.083403              -0.553080   
p77                 0.061592               0.495358               0.129456   
p78                 0.285428               0.316065              -0.294070   

       principal_component_4  principal_component_5  principal_component_6  
p72                -0.376385               0.022153              -0.221094  
p73                 0.253079              -0.005120               0.238797  
p74_A              -0.179086               0.002877              -0.094133  
p74_B               0.440269              -0.384985              -0.543735  
p74_C               0.077061               0.818286               0.088597  
p74_D              -0.400872              -0.417421               0.559154  
p74_E               0.122890              -0.001928               0.075670  
p74_F               0.083413              -0.003927              -0.036812  
p75                -0.055554              -0.015267               0.043286  
p76                 0.223545              -0.031864               0.319573  
p77                 0.362037              -0.061132               0.316432  
p78                -0.442211               0.049237              -0.240563

In [81]:
varianza_explicada

array([0.12741799, 0.11039032, 0.10140412, 0.09962585, 0.09775797,
       0.09734793])

La pregunta 74 respuesta a se asocia al primer componente principal, la pregunta 75 al segundo componente principal, la pregunta 76 se asocia al tercer componente principal, las preguntas 74 respuesta b y 78 se asocian al cuarto componente principal, la pregunta 74 respuesta c se asocia al quinto componente principal y la pregunta 74 respuesta d se asocia al sexto componente principal. El porcentaje de varianza explicada de cada componente es de 12.74%, 11.04%, 10.14%, 9.96%, 9.78% y 9.73%, respectivamente.

- Como se observa a continuación, seis eigenvalues correspondientes a los seis componentes principales son mayores a 1, por lo que se usan esos seis en la regresión.

In [82]:
eigenvalues

array([1.52901864, 1.32468625, 1.21685165, 1.19551239, 1.17309772,
       1.16817727])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [83]:
componentesprincipales_seccion_7_2 = pd.DataFrame(componentesprincipales, columns=['sec_7_2_cp_1', 'sec_7_2_cp_2', 'sec_7_2_cp_3', 'sec_7_2_cp_4', 'sec_7_2_cp_5', 'sec_7_2_cp_6'])
componentesprincipales_seccion_7_2

sec_7_2_cp_1  sec_7_2_cp_2  sec_7_2_cp_3  sec_7_2_cp_4  sec_7_2_cp_5  \
0          -1.692423     -0.389568      0.290743     -0.226833      0.027104   
1          -1.746360     -0.219468     -0.837257      0.229085     -0.037882   
2          -1.199705      0.385863     -1.400462     -0.617843      0.056417   
3          -1.928579     -0.421245     -0.649522      0.511394     -0.069315   
4           0.911272     -0.940503     -1.092544     -1.083894     -1.138179   
...              ...           ...           ...           ...           ...   
560621      1.464730     -0.549933      1.329617     -2.174435     -1.060355   
560622      0.911272     -0.940503     -1.092544     -1.083894     -1.138179   
560623      1.075952     -0.461068     -1.554212     -0.035900      2.374544   
560624      1.075952     -0.461068     -1.554212     -0.035900      2.374544   
560625      1.299682      2.780455     -0.336412      2.045441     -0.290182   

        sec_7_2_cp_6  
0          -0.307604  
1           0.344161  
2          -0.116568  
3           0.497738  
4           1.778152  
...              ...  
560621      0.527574  
560622      1.778152  
560623      0.325899  
560624      0.325899  
560625      2.002794  

[560626 rows x 6 columns]

##### Sección 8. Entorno escolar 

- 8.1. Entorno escolar

In [84]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_8_1 = df.loc[:, ['p79', 'p80', 'p81', 'p82', 'p83', 'p84']] ###Se seleccionan las preguntas que conforman esta sección

In [85]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_8_1 = scaler.fit_transform(seccion_8_1)

In [86]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_8_1)

In [87]:
componentes = pd.DataFrame(pca.components_.T, index=['p79', 'p80', 'p81', 'p82', 'p83', 'p84'])

In [88]:
####Se eligieron cuatro componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2', 2:'principal_component_3', 3:'principal_component_4'})
componentes

principal_component_1  principal_component_2  principal_component_3  \
p79              -0.558616               0.226099               0.236934   
p80              -0.544775               0.290390               0.257000   
p81               0.091670               0.775381               0.045641   
p82              -0.354223              -0.488257               0.355332   
p83              -0.261403               0.075687              -0.727776   
p84              -0.434695              -0.138611              -0.468839   

     principal_component_4  
p79              -0.310093  
p80               0.290081  
p81               0.051692  
p82               0.368839  
p83               0.569519  
p84              -0.597184

In [89]:
varianza_explicada

array([0.22710596, 0.18170171, 0.17667543, 0.14588125])

La pregunta 79 se asocia al primer componente principal, esa misma pregunta se asocia al segundo componente principal, la pregunta 83 se asocia al tercer componente principal y la pregunta 84 se asocia al cuarto componente principal. El porcentaje de varianza explicada para cada uno de los componentes principales es de 22.71%, 18.17%, 17.66% y 14.59%, respectivamente. 

- Como se muestra a continuación, sólo tres eigenvalues son mayores a 1, por lo que ese criterio se usa para seleccionar a los tres componentes principales que se utilizan en la regresión.

In [90]:
eigenvalues

array([1.36263819, 1.0902122 , 1.06005444, 0.87528909])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [91]:
componentesprincipales_seccion_8_1 = pd.DataFrame(componentesprincipales)
componentesprincipales_seccion_8_1_regress = componentesprincipales_seccion_8_1.loc[:, [0, 1, 2]]

In [92]:
componentesprincipales_seccion_8_1_regress = componentesprincipales_seccion_8_1_regress.rename(columns={0:'sec_8_1_cp_1', 1:'sec_8_1_cp_2', 2:'sec_8_1_cp_3'})
componentesprincipales_seccion_8_1_regress

sec_8_1_cp_1  sec_8_1_cp_2  sec_8_1_cp_3
0          -0.646409      0.018108     -0.756440
1           1.042320     -1.601128     -1.278587
2           1.629055      0.272453     -1.641406
3           1.666530     -1.781863      0.459283
4           0.770477     -1.419187      0.839339
...              ...           ...           ...
560621     -0.646409      0.018108     -0.756440
560622     -1.515797      1.200598     -0.620821
560623      1.631840     -1.064143     -0.007648
560624     -1.082745     -0.583331     -0.318738
560625     -0.210073      0.619548     -1.194141

[560626 rows x 3 columns]

- 8.2. Maestros  de Español (literatura, lectura y redacción, etcétera)

In [93]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_8_2 = df.loc[:, ['p85', 'p86', 'p87', 'p88', 'p89']] ###Se seleccionan las preguntas que conforman esta sección

In [94]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_8_2 = scaler.fit_transform(seccion_8_2)

In [95]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_8_2)

In [96]:
componentes = pd.DataFrame(pca.components_.T, index=['p85', 'p86', 'p87', 'p88', 'p89'])

In [97]:
####Se eligieron dos componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2'})
componentes

principal_component_1  principal_component_2
p85              -0.455160               0.456651
p86              -0.448956               0.524851
p87              -0.469865              -0.093028
p88              -0.452311              -0.303598
p89              -0.407319              -0.644341

In [98]:
varianza_explicada

array([0.59393346, 0.13322657])

Para el primer componente principal la pregunta 87 esta asociada a este, mientras que la pregunta 89 esta asociada al componente principal 2. El porcentaje de varianza explicada para cada uno de los componentes es de 59.39% y de 13.32%, respectivamente.

- Se observa a continuación que sólo un eigenvalue es mayor a 1, por lo que con este criterio se decide elegir el primer componente principal que será usado en la regresión.

In [99]:
eigenvalues

array([2.96967259, 0.66613403])

In [100]:
componentesprincipales_seccion_8_2 = pd.DataFrame(componentesprincipales)
componentesprincipales_seccion_8_2_regress = componentesprincipales_seccion_8_2.loc[:, [0]]

In [101]:
componentesprincipales_seccion_8_2_regress = componentesprincipales_seccion_8_2_regress.rename(columns={0:'sec_8_2_cp_1'})

In [102]:
componentesprincipales_seccion_8_2_regress

sec_8_2_cp_1
0          -0.089979
1           1.455569
2           0.974236
3           3.066611
4          -0.122726
...              ...
560621     -0.122726
560622     -0.045681
560623      3.675474
560624     -2.262203
560625     -2.827346

[560626 rows x 1 columns]

- 8.3. Maestros  de Matemáticas

In [103]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_8_3 = df.loc[:, ['p90', 'p91', 'p92', 'p93', 'p94', 'p95', 'p96', 'p97', 'p98', 'p99', 'p100', 'p101', 'p102', 'p103', 
                        'p104','p105','p106', 'p107', 'p108', 'p109', 'p110', 'p111']] ###Se seleccionan las preguntas que conforman esta sección

In [104]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_8_3 = scaler.fit_transform(seccion_8_3)

In [105]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_8_3)

In [106]:
componentes = pd.DataFrame(pca.components_.T, index=['p90', 'p91', 'p92', 'p93', 'p94', 'p95', 'p96', 'p97', 'p98', 'p99', 'p100', 'p101', 'p102', 'p103', 
                        'p104','p105','p106', 'p107', 'p108', 'p109', 'p110', 'p111'])

In [107]:
####Se eligieron cinco componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2', 2:'principal_component_3', 3:'principal_component_4',
                                         4:'principal_component_5'})
componentes

principal_component_1  principal_component_2  principal_component_3  \
p90                0.097555               0.198541               0.628208   
p91                0.100817               0.181991               0.635644   
p92               -0.182933              -0.193959               0.260685   
p93               -0.207964              -0.289249               0.238520   
p94               -0.234642              -0.246913               0.193538   
p95               -0.235859              -0.089995               0.153864   
p96               -0.243058              -0.130442               0.031113   
p97               -0.218914              -0.088630               0.017895   
p98               -0.240096               0.024432               0.003571   
p99               -0.257255              -0.143139               0.011624   
p100              -0.252915              -0.155233               0.003655   
p101              -0.236522              -0.182406               0.010066   
p102              -0.177403               0.312452              -0.006600   
p103              -0.158893               0.370534               0.008527   
p104              -0.218599               0.083975              -0.041183   
p105              -0.233184               0.067400              -0.043400   
p106              -0.225164               0.191138              -0.044251   
p107              -0.234029               0.106255              -0.047309   
p108              -0.247251               0.094563              -0.040915   
p109              -0.246251               0.006509              -0.045330   
p110              -0.172471               0.387379              -0.033569   
p111              -0.162228               0.424733              -0.032149   

      principal_component_4  principal_component_5  
p90                0.218375               0.009891  
p91                0.225183               0.006874  
p92               -0.354973               0.116360  
p93               -0.309699               0.051906  
p94               -0.322182               0.055457  
p95               -0.258282               0.126591  
p96                0.154672              -0.167977  
p97                0.240642              -0.333729  
p98                0.169914              -0.175234  
p99                0.161650              -0.139111  
p100               0.093345              -0.170487  
p101               0.169038              -0.229072  
p102              -0.163072              -0.431722  
p103              -0.205982              -0.392276  
p104               0.207238               0.159178  
p105               0.221858               0.207818  
p106               0.082609               0.292442  
p107               0.179975               0.278253  
p108               0.096895               0.264001  
p109               0.127754               0.193527  
p110              -0.221305               0.105619  
p111              -0.244875               0.057337

In [108]:
##Se convierte a valor absoluto el dataframe para poder explicar qué preguntas se asocian a cada componente.
componentes = componentes.abs()

In [109]:
componentes.idxmax(axis=0) ##Se obtienen las preguntas que en valor absoluto se asocian más a cada componente.

principal_component_1     p99
principal_component_2    p111
principal_component_3     p91
principal_component_4     p92
principal_component_5    p102
dtype: object

In [110]:
####Se obtiene la varianza explicada por cada componente principal
varianza_explicada = pd.DataFrame(varianza_explicada, index = componentes.columns)
varianza_explicada['var_exp'] = varianza_explicada[0] * 100
varianza_explicada = varianza_explicada.rename(columns = {'var_exp':'porcentaje_varianza_explicada'})
varianza_explicada = varianza_explicada.drop([0], axis = 1)

In [111]:
varianza_explicada ##Porcentaje de varianza explicada por cada componente principal

porcentaje_varianza_explicada
principal_component_1                      37.257887
principal_component_2                      10.006046
principal_component_3                       6.066685
principal_component_4                       5.109716
principal_component_5                       4.710067

- Se observa a continuación que todos los eigenvalues son mayores a 1, por lo que se utilizan todos los componentes en la regresión.

In [112]:
eigenvalues

array([8.19674973, 2.20133406, 1.33467312, 1.12413963, 1.03621656])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [113]:
componentesprincipales_seccion_8_3 = pd.DataFrame(componentesprincipales, columns=['sec_8_3_cp_1', 'sec_8_3_cp_2', 'sec_8_3_cp_3', 'sec_8_3_cp_4', 'sec_8_3_cp_5'])
componentesprincipales_seccion_8_3

sec_8_3_cp_1  sec_8_3_cp_2  sec_8_3_cp_3  sec_8_3_cp_4  sec_8_3_cp_5
0          -3.628529      0.496569     -0.588231     -0.814625      2.277918
1           2.930330     -1.429774     -0.322693      0.776891     -0.983007
2          -0.224740     -0.550760      0.918346     -1.598281     -0.872882
3           4.081648     -0.218684      0.314330      0.374673     -1.079917
4          -0.618941      0.943795     -1.049338     -0.765747      0.896556
...              ...           ...           ...           ...           ...
560621      1.554335      1.646757      0.579080     -0.058506      0.743918
560622      0.991111     -1.906667      0.718886     -0.932218     -0.079320
560623      2.218145     -1.050477      1.320338     -0.488650     -1.415720
560624      4.490182      0.008542     -1.575782     -0.411108      0.048593
560625     -3.304853      1.208908      2.321712      1.038138     -0.317166

[560626 rows x 5 columns]

##### Sección 9. Entorno escolar 

In [114]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_9 = df.loc[:, ['p112', 'p113', 'p114', 'p115', 'p116', 'p117','p118','p119', 'p120', 'p121', 'p122', 'p123',
                       'p124', 'p125','p126','p127', 'p128', 'p129', 'p130', 'p131']] ###Se seleccionan las preguntas que conforman esta sección

In [115]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_9 = scaler.fit_transform(seccion_9)

In [116]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_9)

In [117]:
componentes = pd.DataFrame(pca.components_.T, index=['p112', 'p113', 'p114', 'p115', 'p116', 'p117','p118','p119', 'p120', 'p121', 'p122', 'p123',
                       'p124', 'p125','p126','p127', 'p128', 'p129', 'p130', 'p131'])

In [118]:
####Se eligen 4 componentes principal, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2', 2:'principal_component_3', 
                                          3:'principal_component_4'})
componentes

principal_component_1  principal_component_2  principal_component_3  \
p112               0.247017               0.285724              -0.198749   
p113               0.268903               0.250032              -0.159107   
p114               0.264026               0.251962              -0.166412   
p115               0.263305               0.233802              -0.115502   
p116               0.251102               0.275569              -0.195127   
p117               0.242388               0.289227              -0.205163   
p118              -0.168379               0.112452              -0.059052   
p119               0.248989              -0.036660               0.246381   
p120               0.242579              -0.077374               0.290419   
p121               0.224353              -0.098107               0.272991   
p122               0.211144              -0.071175               0.226754   
p123               0.254986              -0.053447               0.278689   
p124               0.280179              -0.046518               0.290947   
p125               0.260131              -0.047952               0.273204   
p126              -0.117898               0.190175               0.113041   
p127              -0.164648               0.252661               0.177608   
p128              -0.176128               0.326776               0.243392   
p129              -0.169041               0.325632               0.225503   
p130              -0.160941               0.333947               0.273976   
p131              -0.155018               0.334330               0.270321   

      principal_component_4  
p112               0.043842  
p113              -0.071136  
p114              -0.058085  
p115              -0.057208  
p116               0.086365  
p117               0.095206  
p118              -0.525175  
p119              -0.327961  
p120              -0.173638  
p121               0.244805  
p122               0.435163  
p123              -0.082473  
p124              -0.213909  
p125              -0.053913  
p126              -0.452977  
p127               0.133841  
p128               0.096550  
p129               0.112854  
p130               0.062369  
p131               0.050193

In [119]:
varianza_explicada

array([0.28113402, 0.16371373, 0.12106708, 0.04717534])

La pregunta 124 se asocia al primer componente principal, las preguntas 130 y 131 se asocian al segundo componente principal, las preguntas 120 y 124 se asocian al tercer componente principal y la pregunta 118 se asocia al cuarto componente principal. Los porcentajes de varianza explicada de cada componente es de 28.11%, 16.37%, 12.11% y 4.72%, respectivamente.

- Se observa a continuación que tres eigenvalue son mayores a 1, por lo que nos quedamos con esos tres para incorporarlos en la regresión.

In [120]:
eigenvalues

array([5.62269039, 3.27428035, 2.42134596, 0.94350858])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [121]:
componentesprincipales_seccion_9 = pd.DataFrame(componentesprincipales)
componentesprincipales_seccion_9_regress = componentesprincipales_seccion_9.loc[:, [0, 1, 2]]

In [122]:
componentesprincipales_seccion_9_regress = componentesprincipales_seccion_9_regress.rename(columns={0:'sec_9_cp_1', 1:'sec_9_cp_2', 2:'sec_9_cp_3' })                                                                           

In [123]:
componentesprincipales_seccion_9_regress

sec_9_cp_1  sec_9_cp_2  sec_9_cp_3
0        -1.007938    1.674925    2.033414
1         0.301634   -1.639380    0.733296
2         2.382538   -1.359657   -1.159584
3        -0.223632    0.026949   -0.358396
4        -0.219535   -1.224178   -0.397484
...            ...         ...         ...
560621    0.113819   -0.927146   -1.117167
560622   -0.735122    1.061960   -0.669647
560623    0.473479   -1.013629    2.070286
560624    1.461200    0.442900    1.430658
560625   -2.480767    1.737861    1.810946

[560626 rows x 3 columns]

##### Sección 10. Entorno familiar 

- 10.1. Entorno familiar

In [124]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_10_1 = df.loc[:, ['p132', 'p133', 'p134', 'p135', 'p136', 'p137', 'p138', 'p139', 'p140']] ###Se seleccionan las preguntas que conforman esta sección

In [125]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_10_1 = scaler.fit_transform(seccion_10_1)

In [126]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_10_1)

In [127]:
componentes = pd.DataFrame(pca.components_.T, index=['p132', 'p133', 'p134', 'p135', 'p136', 'p137', 'p138', 'p139', 'p140'])

In [128]:
####Se eligieron 4 componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2', 2:'principal_component_3', 
                                          3:'principal_component_4'})
componentes

principal_component_1  principal_component_2  principal_component_3  \
p132              -0.186674              -0.156195              -0.617559   
p133              -0.135244              -0.200085              -0.666028   
p134              -0.305512              -0.078426              -0.131623   
p135              -0.489529              -0.230987               0.203623   
p136              -0.431245              -0.332629               0.201464   
p137              -0.441717              -0.162574               0.223563   
p138              -0.281131               0.550831              -0.077297   
p139              -0.277631               0.550454              -0.126173   
p140              -0.276921               0.364112               0.061461   

      principal_component_4  
p132              -0.318670  
p133              -0.024132  
p134               0.844588  
p135              -0.119711  
p136               0.069939  
p137              -0.274824  
p138               0.061996  
p139               0.059801  
p140              -0.286998

In [129]:
varianza_explicada

array([0.26627842, 0.19714304, 0.13316605, 0.09865021])

La pregunta asociada al primer componente principal es la 135, las preguntas asociadas al segundo componente principal son la 138 y 139, la pregunta 133 se encuentra asociada al tercer componente principal y la pregunta 134 se encuentra asociada al cuarto componente principal. El porcentaje de varianza explicada para cada componente principal es de 26.63%, 19.71%, 13.32% y de 9.87% respectivamente.

- Se observa que únicamente el último eigenvalue es menor a 1, por lo tanto con este criterio se eligen 3 componentes principales que se incorporan en la regresión. 

In [130]:
eigenvalues

array([2.39651002, 1.77429053, 1.19849656, 0.88785345])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [131]:
componentesprincipales_seccion_10_1 = pd.DataFrame(componentesprincipales)
componentesprincipales_seccion_10_1_regress = componentesprincipales_seccion_10_1.loc[:, [0, 1, 2]]

In [132]:
componentesprincipales_seccion_10_1_regress = componentesprincipales_seccion_10_1_regress.rename(columns={0:'sec_10_1_cp_1', 1:'sec_10_1_cp_2', 2:'sec_10_1_cp_3'})

In [133]:
componentesprincipales_seccion_10_1_regress

sec_10_1_cp_1  sec_10_1_cp_2  sec_10_1_cp_3
0           -1.310197      -1.086806       0.027878
1           -0.761518       1.965701      -0.450662
2           -1.204937      -0.170071      -0.351054
3           -0.894177       1.179568       1.859064
4            0.792073       0.234584      -0.519467
...               ...            ...            ...
560621       0.274730      -0.330014      -0.772726
560622       0.695355      -1.963378      -0.516205
560623      -0.633179       1.367067      -0.860962
560624      -1.955804       1.309561      -0.622340
560625      -0.025378      -0.864973      -0.447368

[560626 rows x 3 columns]

- 10.2. En tu casa, ¿cuentas con los siguientes bienes y servicios?

In [134]:
np.random.seed(150) ###Se define la semilla para la reproducibilidad de los resultados

seccion_10_2 = df.loc[:, ['p141', 'p142', 'p143', 'p144', 'p145', 'p146', 'p147', 'p148', 'p149', 'p150', 'p151', 'p152',
                          'p153', 'p154']] ###Se seleccionan las preguntas que conforman esta sección

In [135]:
####Se escalan los datos.
scaler = preprocessing.StandardScaler()

seccion_10_2 = scaler.fit_transform(seccion_10_2)

In [136]:
####Se obtienen los atributos que conforman la función componentes_principales 
pca, varianza_explicada, componentesprincipales, val_sing, pca.components_, eigenvalues = PCA_from_sklearn.componentes_principales(seccion_10_2)

In [137]:
componentes = pd.DataFrame(pca.components_.T, index=[['p141', 'p142', 'p143', 'p144', 'p145', 'p146', 'p147', 'p148', 'p149', 'p150', 'p151', 'p152',
                          'p153', 'p154']])

In [138]:
####Se eligieron 5 componentes principales, por lo que a continuación se muestran los ejes principales que representan las direcciones de máxima varianza en los datos
componentes = componentes.rename(columns={'index': 'preguntas', 0:'principal_component_1', 1:'principal_component_2', 2:'principal_component_3', 3:'principal_component_4',
                                         4:'principal_component_5'})
componentes

principal_component_1  principal_component_2  principal_component_3  \
p141               0.165506               0.275201               0.406487   
p142               0.236156               0.070485               0.214985   
p143               0.257840               0.345178              -0.194953   
p144              -0.019400               0.445835              -0.573005   
p145              -0.276585               0.188656              -0.187063   
p146              -0.257842               0.339219               0.253437   
p147              -0.194256               0.346358               0.351277   
p148              -0.305824               0.244234               0.035695   
p149              -0.209090               0.199777               0.309703   
p150              -0.323477               0.238689              -0.187301   
p151               0.282028               0.287211              -0.014710   
p152               0.352063               0.147396               0.200110   
p153               0.341620               0.195144              -0.162937   
p154               0.327394               0.178675               0.022586   

      principal_component_4  principal_component_5  
p141              -0.436609              -0.352655  
p142              -0.025947              -0.592200  
p143              -0.313807              -0.065114  
p144              -0.388330               0.182194  
p145               0.341029              -0.426664  
p146               0.081376               0.115743  
p147               0.181328               0.294646  
p148              -0.032285              -0.061604  
p149              -0.150166               0.247152  
p150               0.216740              -0.312857  
p151               0.423027              -0.000302  
p152               0.114743               0.138941  
p153               0.329615               0.085202  
p154               0.179691               0.126179

In [139]:
varianza_explicada

array([0.30711338, 0.09780948, 0.08020446, 0.07305564, 0.06357539])

La pregunta 152 se asocia al primer componente principal, la pregunta 144 se asocia al segundo componente principal y al tercer componente principal, la pregunta 141 al cuarto componente principal, la pregunta 142 se asocia al quinto componente principal. El porcentaje de varianza explicada para cada componente principal es de 30.71%, 9.79%, 8.02%, 7.31% y 6.36%, respectivamente. 

- Se observa que 4 eigenvalues son mayores a 1, por lo que los cuatro componentes principales se incorporarán en la regresión.

In [140]:
eigenvalues

array([4.29959503, 1.3693351 , 1.12286443, 1.02278072, 0.89005698])

- Se obtienen los componentes principales que son los que se meteran a la regresión.

In [141]:
componentesprincipales_seccion_10_2 = pd.DataFrame(componentesprincipales)
componentesprincipales_seccion_10_2_regress = componentesprincipales_seccion_10_2.loc[:, [0, 1, 2, 3]]

In [142]:
componentesprincipales_seccion_10_2_regress = componentesprincipales_seccion_10_2_regress.rename(columns={0:'sec_10_2_cp_1', 1:'sec_10_2_cp_2', 2:'sec_10_2_cp_3', 
                                                                                                          3:'sec_10_2_cp_4'})

In [143]:
componentesprincipales_seccion_10_2_regress

sec_10_2_cp_1  sec_10_2_cp_2  sec_10_2_cp_3  sec_10_2_cp_4
0           -0.472417      -0.307271      -2.021332       1.726780
1            1.520469       0.580579      -2.225639       0.677854
2            1.061461      -0.697975      -0.824011      -0.564084
3           -0.579596      -2.437395       2.029714      -0.976424
4            1.607332      -0.691650      -0.699982      -0.110107
...               ...            ...            ...            ...
560621      -2.399400      -0.846558      -0.319832      -0.879166
560622      -0.043987       1.429863      -2.542011       1.606282
560623       1.049268      -0.751805      -0.200687      -1.095462
560624       0.545408      -0.194436       0.601427      -2.210730
560625      -0.112023      -1.543342       0.879039      -0.872082

[560626 rows x 4 columns]

##### A continuación se muestra un Dataframe que contiene los componentes principales seleccionados mediante el porcentaje de varianza explicada.

In [144]:
componentes_principales_var_exp = pd.concat([componentesprincipales_seccion_1, componentesprincipales_seccion_2, componentesprincipales_seccion_3, 
                                            componentesprincipales_seccion_4_1, componentesprincipales_seccion_5_1, componentesprincipales_seccion_5_2,
                                            componentesprincipales_seccion_6, componentesprincipales_seccion_7_1, componentesprincipales_seccion_7_2, componentesprincipales_seccion_8_1,
                                            componentesprincipales_seccion_8_2, componentesprincipales_seccion_8_3, componentesprincipales_seccion_9, 
                                             componentesprincipales_seccion_10_1, componentesprincipales_seccion_10_2], axis = 1)

In [145]:
componentes_principales_var_exp.shape ###Contiene 65 variables.

(560626, 65)

##### En el siguiente Dataframe se presentan los componentes principales seleccionados mediante el criterio de eigenvalue.

In [146]:
componentes_principales_regress = pd.concat([componentesprincipales_seccion_1, componentesprincipales_seccion_2, componentesprincipales_seccion_3,
                                             componentesprincipales_seccion_4_1_regress, componentesprincipales_seccion_5_1, 
                                             componentesprincipales_seccion_5_2_regress, componentesprincipales_seccion_6_regress, 
                                             componentesprincipales_seccion_7_1, componentesprincipales_seccion_7_2,
                                            componentesprincipales_seccion_8_1_regress, componentesprincipales_seccion_8_2_regress,
                                            componentesprincipales_seccion_8_3, componentesprincipales_seccion_9_regress, componentesprincipales_seccion_10_1_regress,
                                            componentesprincipales_seccion_10_2_regress], axis = 1)

In [147]:
componentes_principales_regress.shape ###Contiene 53 variables.

(560626, 53)

In [148]:
componentes_principales_var_exp.to_csv('../../results/data_results/componentes_principales_var_exp.csv', index=False)

In [149]:
componentes_principales_regress.to_csv('../../results/data_results/componentes_principales_regress.csv', index=False)

In [150]:
#secciones = pd.concat([seccion_1, seccion_2, seccion_3, seccion_4, seccion_5_1, seccion_5_2, seccion_6, seccion_7_1, seccion_7_2, seccion_8_1, seccion_8_2, seccion_8_3,
                       #seccion_9, seccion_10_1, seccion_10_2], axis = 1)

In [151]:
#secciones.head()

In [152]:
#secciones.to_csv('../../results/data_results/secciones.csv', index=False)